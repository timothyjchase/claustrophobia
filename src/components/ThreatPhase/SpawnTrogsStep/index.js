import React, { Component } from 'react'
import PropTypes from 'prop-types'
import renderHTML from 'react-render-html'
import { Button, Input, Message } from 'semantic-ui-react'

class SpawnTrogsStep extends Component {
  constructor(props) {
    super(props)
    const { legalPlacement, threatDice } = props
    const maxTrogs = legalPlacement ? threatDice : Math.min(threatDice, 3)
    this.state = { trogsAdded: maxTrogs, maxTrogs }
  }

  render = () => {
    const {
      scenario,
      legalPlacement,
      completeThreatSpawnTrogsStep,
    } = this.props
    const { trogsAdded, maxTrogs } = this.state
    const spawnLocation =
      scenario.trogsSpawnLocation ||
      '<li>the <strong>largest group</strong> of Human warriors</li>'

    return (
      <div>
        {!legalPlacement && (
          <Message negative>
            <Message.Header>A Taste for Blood</Message.Header>
            <Message.Content>
              Human warriors cannot prevent the appearance of Troglodytes.
            </Message.Content>
          </Message>
        )}
        <p>
          Add up to
          <strong>{` ${maxTrogs} `}</strong>
          Troglogytes to tiles closest to:
        </p>
        <ul>{renderHTML(spawnLocation)}</ul>
        {legalPlacement && (
          <p>
            <i>Reminder: Human warriors block tiles with unexplored openings</i>
          </p>
        )}
        <Input
          label="Troglogytes added"
          placeholder="0"
          type="number"
          min={0}
          max={6}
          value={trogsAdded}
          onChange={(event, { value }) => this.setState({ trogsAdded: value })}
          fluid
        />
        <br />
        <Button.Group vertical fluid>
          <Button
            primary
            disabled={!trogsAdded}
            icon="play"
            content="Next"
            onClick={() =>
              completeThreatSpawnTrogsStep(parseInt(trogsAdded, 10))
            }
          />
        </Button.Group>
      </div>
    )
  }
}

SpawnTrogsStep.propTypes = {
  scenario: PropTypes.object.isRequired,
  legalPlacement: PropTypes.bool.isRequired,
  threatDice: PropTypes.number.isRequired,
  completeThreatSpawnTrogsStep: PropTypes.func.isRequired,
}

export default SpawnTrogsStep
