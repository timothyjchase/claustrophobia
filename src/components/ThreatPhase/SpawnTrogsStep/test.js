import React from 'react'
import SpawnTrogsStep from './index'

describe('<SpawnTrogsStep />', () => {
  describe('with location, legal placement, threat 6', () => {
    const props = {
      scenario: { trogsSpawnLocation: 'custom location' },
      legalPlacement: true,
      threatDice: 6,
      completeThreatSpawnTrogsStep: jest.fn(),
    }
    const component = shallow(<SpawnTrogsStep {...props} />)

    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
    it('have max trogs = 6', () => {
      expect(component.state().maxTrogs).toEqual(6)
    })
    it('default trogs added = 6', () => {
      expect(component.state().trogsAdded).toEqual(6)
    })
  })
  describe('with no location, no legal placement, threat 2', () => {
    const props = {
      scenario: {},
      legalPlacement: false,
      threatDice: 2,
      completeThreatSpawnTrogsStep: jest.fn(),
    }
    const component = shallow(<SpawnTrogsStep {...props} />)

    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
    it('have max trogs = 2', () => {
      expect(component.state().maxTrogs).toEqual(2)
    })
    it('default trogs added = 2', () => {
      expect(component.state().trogsAdded).toEqual(2)
    })
  })
  describe('with no location, no legal placement, threat 4', () => {
    const props = {
      scenario: {},
      legalPlacement: false,
      threatDice: 4,
      completeThreatSpawnTrogsStep: jest.fn(),
    }
    const component = shallow(<SpawnTrogsStep {...props} />)

    it('have max trogs = 3', () => {
      expect(component.state().maxTrogs).toEqual(3)
    })
    it('default trogs added = 3', () => {
      expect(component.state().trogsAdded).toEqual(3)
    })
  })
})
