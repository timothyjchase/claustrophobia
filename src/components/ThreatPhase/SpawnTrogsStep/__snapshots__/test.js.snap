// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpawnTrogsStep /> with location, legal placement, threat 6 should match snapshot 1`] = `
<div>
  <p>
    Add up to
    <strong>
       6 
    </strong>
    Troglogytes to tiles closest to:
  </p>
  <ul>
    custom location
  </ul>
  <p>
    <i>
      Reminder: Human warriors block tiles with unexplored openings
    </i>
  </p>
  <Input
    fluid={true}
    label="Troglogytes added"
    max={6}
    min={0}
    onChange={[Function]}
    placeholder="0"
    type="number"
    value={6}
  />
  <br />
  <ButtonGroup
    fluid={true}
    vertical={true}
  >
    <Button
      as="button"
      content="Next"
      disabled={false}
      icon="play"
      onClick={[Function]}
      primary={true}
      role="button"
    />
  </ButtonGroup>
</div>
`;

exports[`<SpawnTrogsStep /> with no location, no legal placement, threat 2 should match snapshot 1`] = `
<div>
  <Message
    negative={true}
  >
    <MessageHeader>
      A Taste for Blood
    </MessageHeader>
    <MessageContent>
      Human warriors cannot prevent the appearance of Troglodytes.
    </MessageContent>
  </Message>
  <p>
    Add up to
    <strong>
       2 
    </strong>
    Troglogytes to tiles closest to:
  </p>
  <ul>
    <li
      key="0"
    >
      the 
      <strong
        key="1"
      >
        largest group
      </strong>
       of Human warriors
    </li>
  </ul>
  <Input
    fluid={true}
    label="Troglogytes added"
    max={6}
    min={0}
    onChange={[Function]}
    placeholder="0"
    type="number"
    value={2}
  />
  <br />
  <ButtonGroup
    fluid={true}
    vertical={true}
  >
    <Button
      as="button"
      content="Next"
      disabled={false}
      icon="play"
      onClick={[Function]}
      primary={true}
      role="button"
    />
  </ButtonGroup>
</div>
`;
